### PR TITLE
solving #28 implementing resolver for schemas with tests

### LINKED ISSUE #28: Implemen ResolverClass

### PR DESCRIPTION
**Pull request by jp06upel**
Closes #28

### CODE REVIEW

File: src/test/java/org/jqassistant/plugin/openapi/parser/ResolverTest.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+            assertThat(e).isNotNull();
+        }
+
+        try {
+            resolver.createIfAbsent("name");
+            Assertions.fail("resolved invalid reference");
+        } catch (Exception e){
+            assertThat(e).isNotNull();
+        }
+
```
Comment (UzeyirMammadov) [inline src/test/java/org/jqassistant/plugin/openapi/parser/ResolverTest.java:78]: Maybe we can use assertThrows() to reduce the number of try-catch blocks.
https://stackoverflow.com/questions/40268446/junit-5-how-to-assert-an-exception-is-thrown

**Review by UzeyirMammadov — COMMENTED** [review]
Review done. Please check my comments.

File: src/main/java/org/jqassistant/plugin/openapi/impl/util/Resolver.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        }
+    }
+
+    /**
+     * Parses ref string to object name
+     *
+     * @param ref ref string in form of "#/components/schemas/example"
+     * @return the object name
+     */
+    private static String refToName(String ref){
```
[the remark this answers is not in the document]
Comment (jp06upel) [reply src/main/java/org/jqassistant/plugin/openapi/impl/util/Resolver.java:41]: first check is length check, so no null pointer should occur, but can do an explicit check if prefered.

File: src/test/java/org/jqassistant/plugin/openapi/parser/ResolverTest.java (hunk 1 above)
Comment (jp06upel) [reply src/test/java/org/jqassistant/plugin/openapi/parser/ResolverTest.java:78]: sounds useful! will switch

File: src/main/java/org/jqassistant/plugin/openapi/impl/util/Resolver.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (UzeyirMammadov) [reply src/main/java/org/jqassistant/plugin/openapi/impl/util/Resolver.java:41]: I mean adding a check before splitting(between lines 41 and 42). Length check happens after.

File: src/main/java/org/jqassistant/plugin/openapi/impl/util/Resolver.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        String[] splits = ref.split("/");
+        if (splits.length != 4
+        || !splits[0].equals("#")
+        || !splits[1].equals("components")
```
[the remark this answers is not in the document]
Comment (UzeyirMammadov) [reply src/main/java/org/jqassistant/plugin/openapi/impl/util/Resolver.java:45]: What about this? Using constants will make it more readable and understandable. For example, if a person opens the project for the first time he won't understand immediately what "#" is in this context, using a constant instead will make it easier. Besides it easy to make a typo using strings, especially if they are used multiple times.

File: src/test/java/org/jqassistant/plugin/openapi/parser/ResolverTest.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        assertThat(schemaDescriptor.getName()).isEqualTo(objectName);
+    }
+
+    @Test
+    void resolveExistingSchema(){
+        String objectName = "foo";
+        String objectNameEdit = objectName + "not";
+        String objectRef = "#/components/schemas/" + objectName;
+
+        resolver.createIfAbsent(objectRef).setName(objectNameEdit);
```
Comment (CodeOfRob) [inline src/test/java/org/jqassistant/plugin/openapi/parser/ResolverTest.java:45]: setName here unnecessary? resolver already sets name..

File: src/main/java/org/jqassistant/plugin/openapi/impl/util/Resolver.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (CodeOfRob) [reply src/main/java/org/jqassistant/plugin/openapi/impl/util/Resolver.java:41]: would prefer explicit check too

File: src/main/java/org/jqassistant/plugin/openapi/impl/util/Resolver.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (CodeOfRob) [reply src/main/java/org/jqassistant/plugin/openapi/impl/util/Resolver.java:45]: explanation is given by example in functions doc comment, reoccurring usage of these strings are imo also unlikely.
in order to achieve clean code constants like suggested would be good

dont really care which way in this situation

[the remark this answers is not in the document]
Comment (UzeyirMammadov) [reply src/main/java/org/jqassistant/plugin/openapi/impl/util/Resolver.java:45]: Didn't notice the doc comment. But still, it will make it easier to read.

**Comment by jp06upel** [comment]
> Review done. Please check my comments.

should be all done

**Review by CodeOfRob — APPROVED** [review]
good job

**Pull request merged**